File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys
input = sys.stdin.readline

def bin(arr, start, end):
    mid = (end + start) // 2
    cnt = 0
    for i in arr:
        cnt += i // mid
    
    if cnt < n:
        end = mid - 1
        bin(arr, start, end)
    elif cnt >= n:
        result.append(mid)
        start = mid + 1
        if start > end:
            return True
        bin(arr, start, end)
    return True

k , n = map(int, input().split())
data = [int(input()) for _ in range(k)]
longest = max(data)
result = []
bin(data, 1, longest)
print(max(result))

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.IOException;

public class Main {
    static int[] data;

    public static void main(String[] args) throws IOException {
        int K = readInt();
        int N = readInt();
        
        data = new int[K];
        long max = 0;
        for (int i = 0; i < K; i++) {
            data[i] = readInt();
            max = Math.max(max, data[i]);
        }

        max++;
        long min = 0;
        long mid, sum;

        while (min < max) {
            mid = (max + min) / 2;
            sum = 0;

            for (int i = 0; i < K; i++) {
                sum += (data[i] / mid);
            }

            if (sum < N) {
                max = mid;
            } else {
                min = mid + 1;
            }
        }

        System.out.println(min-1);
    }

    private static int readInt() throws IOException {
        int v, t = 0;
        while ((v = System.in.read()) > 32) {
            t = t * 10 + (v - 48);
        }
        if (v == 13) {
            System.in.read();
        }
        return t;
    }
}
